0000-deb-repo-additional-headers.md

```diff
@@ -0,0 +1,227 @@
+
+
+### Key import via the UI
+
+The UI for creating/updating a repository (`Software -> Manage -> Repositories`) will be extended with two additional fields:
+- GPG key URL / Keyserver
+- GPG key ID
+
+These fields should be enabled only if checkbox `Has Signed Metadata` is checked. If present these fields will be saved into `rhnContentSourceGPG`.
+
+When saving the repository the user will be asked to confirm the import of the GPG key. The confirmation is only needed if `Has Signed Metadata` is checked and if the key doesn't already exist in the `$GPG_DIR` keyring.
+
+The key should be imported into `$GPG_DIR`.
+
+### Key import via the API
+
+The API method `channel.software.create_repo()` should be extended with two optional arguments to allow specifying the `GPG key URL / Keyserver` and `GPG key ID`.
+
+An error should be thrown if importing the GPG keys is not successful.
+
+### Key import via `spacewalk-repo-sync`
+
+`spacewalk-repo-sync` must be extended to be able to import the GPG keys configured in `rhnContentSourceGPG`. Similarly to Yum repos, the user should be asked for confirmation when importing the keys.
+
+### Metadata and signature lookup
+
+In order to verify integrity of the `Packages.gz` file Uyuni must first lookup the `Release` file  and verify its signature. This file can be in one of these locations:
+- `http://mirror.example.com/$ARCHIVE_ROOT/dists/$DIST` in case of repos with a `dists` directory when the Uyuni repo URL points to `http://mirror.example.com/$ARCHIVE_ROOT/dists/$DIST/$COMPONENT/binary-$ARCH/`.
+- `http://mirror.example.com/$REPODIR/` in the case of flat repositories
+
+Looking up the `Release` file should be done automatically when syncing of the repo. An error should be shown if it's not possible to locate the `Release` file and its corresponding signature.
+
+### Metadata Verification
+
+Uyuni should verify the integrity of the `Packages.gz|xz` files using the checksums present in the `Release` file.
+
+### Signature verification
+
+If the `InRelease` or `Release.pgg` exist in the repo then Uyuni must verify the `Release` file:
+```
+gpgv --keyring /var/lib/spacewalk/gpgdir/pubring.gpg InRelease
+```
+or
+```
+gpgv --keyring /var/lib/spacewalk/gpgdir/pubring.gpg Release.gpg Release
+```
+
+The signature verification should be done during repo syncing when the metadata is parsed.
+Verification is possible only if the user has configured the GPG signing of the repos to be synced. Otherwise a warning should be printed in the logs about the lack of GPG configuration.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+No drawbacks.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Support only a very limited set of additional fields like `Multi-Arch` discarding all others.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should all the fields be stored ? Should only the fields defined in [1] be stored ? If yes what happens with user degined fields [2]
```

Comment:
How the end-user would be affected if we just store a limited subset (e.g. 'Multi-Arch')?

Reply:
Other fields might be important for the user as well, although for now `Multi-Arch` is critical for now. For start we could just store all fields and add some kind of filtering later if really needed.